[RNMobile] Add custom matcher `toBeVisible`

fluiddot opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Add custom matcher `toBeVisible` in order to provide more accurate assertions, in terms of component rendering, in integration tests.

**NOTE:** This custom matcher is in the process to be included in the `@testing-library/jest-native` package ([PR reference](https://github.com/testing-library/jest-native/pull/73)). Once it's included and a new version released, it's recommended to replace this implementation with the package.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
The matcher is only used for now in one test ([`packages/react-native-editor/src/test/index.test.js`](https://github.com/WordPress/gutenberg/blob/f02d823fc7101906976c4fcf61ad3b8b3c7d8c3f/packages/react-native-editor/src/test/index.test.js#L192)), so we'd only need to check that the test passes either by checking the PR checks (`Unit Tests / Mobile`) or running the command `npm run native test`.

## Screenshots <!-- if applicable -->
N/A

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Enhancement

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [x] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

fluiddot committed:
Add toBeVisible custom jest matcher

fluiddot committed:
Use custom matcher in editor initialization test

fluiddot committed:
Update react native integration test guide

fluiddot committed:
Merge branch 'trunk' into rnmobile/add/to-be-visible-jest-matcher

# Conflicts:
#	packages/react-native-editor/src/test/index.test.js

fluiddot committed:
Add documentation to toBeVisible custom matcher

dcalhoun commented on docs/contributors/code/react-native/integration-test-guide.md:
```diff
@@ -181,13 +181,13 @@ fireEvent( heightSlider, 'valueChange', '50' );
 ## Expect correct element behaviour
 
-After querying elements and firing events, we have to verify that the logic works as expected, for this purpose we can use the same `expect` function from Jest as we use in unit tests.
+After querying elements and firing events, we have to verify that the logic works as expected, for this purpose we can use the same `expect` function from Jest as we use in unit tests. It's recommended to use the custom matcher `toBeVisible` in order to help us assure that the element is defined, is a valid React element and visible.
```
```suggestion
After querying elements and firing events, we must verify that the logic works as expected. For this purpose we can use the same `expect` function from Jest that we use in unit tests. It is recommended to use the custom `toBeVisible` matcher to ensure the element is defined, is a valid React element, and visible.
```

Minor grammar modifications to improve comprehensibility. Feel free to ignore if you disagree.

fluiddot replied:
Thanks for the suggestions, I've applied them in https://github.com/WordPress/gutenberg/pull/38254/commits/023e8872b2dddf296ee7fe55b68f7be82a983196.

dcalhoun commented on docs/contributors/code/react-native/integration-test-guide.md:
```diff
@@ -338,7 +338,7 @@ If you have trouble locating an element’s identifier, you may wish to use Xcod
 ### False positives when omitting `await` before `waitFor` function
 
-Omitting the `await` before a `waitFor` can lead to scenarios where tests pass but are not testing the intended behaviour. For example, if `toBeDefined` is used to assert the result of a call to `waitFor`, the assertion will pass because `waitFor` returns a value, even though it’s not the `ReactTestInstance` we meant to check for.
+Omitting the `await` before a `waitFor` can lead to scenarios where tests pass but are not testing the intended behaviour. For example, if `toBeDefined` is used to assert the result of a call to `waitFor`, the assertion will pass because `waitFor` returns a value, even though it’s not the `ReactTestInstance` we meant to check for. For this reason, it's recommended to use the custom matcher `toBeV ...
```
```suggestion
Omitting the `await` before a `waitFor` can lead to scenarios where tests pass but are not testing the intended behaviour. For example, if `toBeDefined` is used to assert the result of a call to `waitFor`, the assertion will pass because `waitFor` returns a value, even though it is not the `ReactTestInstance` we meant to check for. For this reason, it is recommended to use the custom matcher `toBeVisible` which guards against this type of false positive.
```

Minor grammar modifications to improve comprehensibility. Feel free to ignore if you disagree.

fluiddot replied:
Thanks for the suggestions, I've applied them in https://github.com/WordPress/gutenberg/pull/38254/commits/023e8872b2dddf296ee7fe55b68f7be82a983196.

dcalhoun commented on test/native/custom-matchers/utils.js:
```diff
@@ -0,0 +1,60 @@
+		throw new ReactElementTypeError( element, ...args );
+	}
+}
+
+export function printElement( { type, props } ) {
```
```suggestion

// "printElement" util function is extracted from "@testing-library/jest-native" package.
// Reference: https://github.com/testing-library/jest-native/blob/0dbda7191106f15a57eda1be10eadf95ad8bd81f/src/utils.js#L59-L68
export function printElement( { type, props } ) {
```

For sake of consistency, WDYT about including a similar origin comment here?

fluiddot replied:
Ok, although I'd like to note that the logic is different than the one provided from https://github.com/testing-library/jest-native/blob/0dbda7191106f15a57eda1be10eadf95ad8bd81f/src/utils.js#L59-L68.

dcalhoun replied:
Ah, OK. I overlooked the divergence. Feel free to not included a reference if you'd prefer.

fluiddot replied:
No worries, I finally included it so we have a reference to the original function.

dcalhoun commented on test/native/setup-after-env.js:
```diff
@@ -0,0 +1,9 @@
+/**
+ * Internal dependencies
+ */
+import { toBeVisible } from './custom-matchers/to-be-visible';
```
```suggestion
import { toBeVisible } from './matchers/to-be-visible';
```

Nitpick: IMO the term "custom" is likely unnecessary. 

N.b. this suggestion should not be accepted directly here, as the file paths need to be updated as well.

fluiddot replied:
To be honest, I wasn't sure about whether to add that term. I ended up picking `custom` because I saw references in either the `@testing-library/jest-native` [README](https://github.com/testing-library/jest-native) and in the Jest documentation ([reference](https://jestjs.io/docs/expect#custom-matchers-api)).

In any case, I like the idea of renaming it 👍 .

fluiddot replied:
I applied this suggestion in https://github.com/WordPress/gutenberg/pull/38254/commits/b2cd5b411a4e600a0eba3768dd2ca28300733da6.

dcalhoun reviewed (approved):
🚀 Very happy to see this land. I have a feeling it may be awhile before the contribution to `@testing-library/jest-native` lands, unfortunately. There appears to be little activity in the repository. 

I left a few suggestions, but all of them are largely mere opinions. Feel free to merge as-is. Thanks!

fluiddot committed:
Improve comprehensibility of RN integration test guide

Apply grammar modifications to improve comprehensibility in the React Native integration test guide.

Co-authored-by: David Calhoun <438664+dcalhoun@users.noreply.github.com>

fluiddot committed:
Add reference to the original printElement function

Co-authored-by: David Calhoun <438664+dcalhoun@users.noreply.github.com>

fluiddot committed:
Rename custom Jest matchers folder
